clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
-            }
-            if rust_cc > self.limit.limit() {
-                span_help_and_lint(
-                    cx,
-                    COGNITIVE_COMPLEXITY,
-                    span,
-                    &format!("the function has a cognitive complexity of {}", rust_cc),
-                    "you could split it up into multiple smaller functions",
-                );
-            }
-        }
-    }
-}
+                helper.visit_block(&fn_block);
+
+                let fn_score = helper.score as u64;
+                let score_limit = self.limit.limit();
 
-impl<'a, 'tcx> LateLintPass<'a, 'tcx> for CognitiveComplexity {
-    fn check_fn(
-        &mut self,
-        cx: &LateContext<'a, 'tcx>,
-        _: intravisit::FnKind<'tcx>,
-        _: &'tcx FnDecl,
-        body: &'tcx Body,
-        span: Span,
-        hir_id: HirId,
-    ) {
-        let def_id = cx.tcx.hir().local_def_id_from_hir_id(hir_id);
-        if !cx.tcx.has_attr(def_id, "test") {
-            self.check(cx, body, span);
+                if fn_score > score_limit {
+                    span_help_and_lint(
+                        cx,
+                        COGNITIVE_COMPLEXITY,
+                        item.span,
+                        &format!("the function has a cognitive complexity of {}", fn_score),
+                        "you could split it up into multiple smaller functions",
+                    );
+                }
+            }
         }
     }
 
-    fn enter_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn enter_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.push_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
-    fn exit_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn exit_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.pop_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
 }
 
-struct CCHelper<'a, 'tcx: 'a> {
-    match_arms: u64,
-    divergence: u64,
-    returns: u64,
-    short_circuits: u64, // && and ||
-    cx: &'a LateContext<'a, 'tcx>,
+struct CoCHelper {
+    current_nesting: u64,
+    score: f64,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for CCHelper<'a, 'tcx> {
-    fn visit_expr(&mut self, e: &'tcx Expr) {
-        match e.node {
-            ExprKind::Match(_, ref arms, _) => {
-                walk_expr(self, e);
-                let arms_n: u64 = arms.iter().map(|arm| arm.pats.len() as u64).sum();
-                if arms_n > 1 {
-                    self.match_arms += arms_n - 2;
-                }
+impl CoCHelper {
+    fn push_nesting(&mut self) {
+        self.current_nesting += 1;
+    }
+
+    fn pop_nesting(&mut self) {
+        assert!(self.current_nesting > 0);
+        self.current_nesting -= 1;
+    }
+
+    fn add_to_score(&mut self, amount: u64) {
+        self.score += amount as f64;
+    }
+
+    fn score_nd_structure(&mut self) {
```

Comment:
Add a doc comment explaining what this function does. (and mention `nesting-dependent`)

Reply:
Okay! :)

Reply:
I've documented it now, is that what you had in mind? :)